tests/sim/PeripheralInputCapture.cpp

```diff
@@ -0,0 +1,248 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * PeripheralInputCapture.cpp
+ * The input capture module used with the simulator.
+ * Copyright (C) 2015 Simon Newton
+ */
+
+#include "PeripheralInputCapture.h"
+
+#include <gtest/gtest.h>
+
+#include <vector>
+
+#include "macros.h"
+#include "Simulator.h"
+#include "ola/Callback.h"
+#include "peripheral/tmr/plib_tmr.h"
+
+PeripheralInputCapture::InputCapture::InputCapture(INT_SOURCE source)
+    : enabled(false),
+      mode(IC_INPUT_CAPTURE_DISABLE_MODE),
+      buffer_size(IC_BUFFER_SIZE_16BIT),
+      edge_type(IC_EDGE_FALLING),
+      timer(IC_TIMER_TMR3),
+      events_per_interrupt(IC_INTERRUPT_ON_EVERY_CAPTURE_EVENT),
+      interrupt_source(source),
+      prescale_counter(0),
+      capture_counter(0),
+      got_trigger(false) {
+}
+
+void PeripheralInputCapture::InputCapture::CaptureEvent(uint32_t value) {
+  if (buffer.size() == FIFO_SIZE) {
+    return;
+  }
+  buffer.push_back(value);
+}
+
+PeripheralInputCapture::PeripheralInputCapture(
+    Simulator *simulator,
+    InterruptController *interrupt_controller)
+    : m_simulator(simulator),
+      m_interrupt_controller(interrupt_controller),
+      m_callback(ola::NewCallback(this, &PeripheralInputCapture::Tick)) {
+  m_simulator->AddTask(m_callback.get());
+  std::vector<INT_SOURCE> timer_sources = {
+    INT_SOURCE_INPUT_CAPTURE_1,
+    INT_SOURCE_INPUT_CAPTURE_2,
+    INT_SOURCE_INPUT_CAPTURE_3,
+    INT_SOURCE_INPUT_CAPTURE_4,
+    INT_SOURCE_INPUT_CAPTURE_5,
+  };
+  for (const auto& source : timer_sources) {
+    m_ic.push_back(InputCapture(source));
+  }
+}
+
+PeripheralInputCapture::~PeripheralInputCapture() {
+  m_simulator->RemoveTask(m_callback.get());
+}
+
+void PeripheralInputCapture::Tick() {
+  for (auto& ic : m_ic) {
```

Comment:
Nope, removing the & would be a bug here because then we'd be operating on a copy of the InputCapture, not the original one.